Fixing rename for permitted subclasses.

lahodaj opened the pull request:
Having something like:
```
sealed interface I permits Subtype {}
final class Subtype implements I {}
```

Renaming `Subtype` will not rename the identifier in the `permits` clause. This is because:

a) the write model (`CausalDiff`) does not support `permits`
b) the indexing does not include the permitted subclasses.

This PR is trying to fix both.

lahodaj commented:
(Found a problem in the patch.)

lahodaj commented:
Should be fixed now, and ready for review. Please let me know what you think!

mbien reviewed (approved):
this looks good as far as I can tell (thanks for the comments in CasualDiff).

merge conflict in apichanges.xml needs to be resolved + rebasing would be good to have a fresh CI run before integration.

mbien reviewed (changes requested):
I found a regression right after review :(

if you rename the interface, it will remove the permits section, this does not happen in NB 24

lahodaj commented:
FWIW: the pre-existing methods now are a bit problematic, as if some code takes a `ClassTree`, and creates a new one from it using the old `TreeMaker.Class/Interface` method with permitted subclasses, the permitted subclasses will be lost. I don't see a good way around that.

So, I've tried to go through the whole NetBeans (production code), and fix what I could. Will push that shortly. I think I still need to make at least one pass through it, maybe add some tests, and maybe add some automated way to prevent use of the pre-existing methods. Will try to look at that as I can.

lahodaj commented:
> @lahodaj Unfortunately I didn't get to taking a look why tests are failing and tomorrow is freeze - probably better to move this to NB 26?

I apparently broke something in the `CasualDiff`. I hope I'll cycle back to this sometime soon.

lahodaj commented:
I think this is ready for review. I've updated, squashed, went through the changes.

The "old" `TreeMaker.Class/Interface` methods will create the AST trees without any permitted subclasses, which could lead to a loss of those permitted subclasses. I don't think there's a reasonable way around that in the generator - the generator needs to follow what is in the new trees. So these old methods are now deprecated and there's an AP that prevents use of the deprecated method inside NB sources.

mbien commented on java/lib.nbjavac/src/org/netbeans/lib/nbjavac/CheckProcessor.java:
```diff
@@ -0,0 +1,88 @@
+                    ((QualifiedNameable) el.getEnclosingElement()).getQualifiedName().contentEquals("org.netbeans.api.java.source.TreeMaker") &&
+                    (el.getSimpleName().contentEquals("Class") || el.getSimpleName().contentEquals("Interface")) &&
+                    processingEnv.getElementUtils().isDeprecated(el)) {
+                    trees.printMessage(Diagnostic.Kind.ERROR, "Use of the deprecated TreeMaker.Class/Interface method is not permitted inside the NetBeans sources.", node, selectPath.getCompilationUnit());
```
this is interesting. Will this run while nb-javac is used within the NB editor too or only during build?

I suppose this is picked up when `-proc:full` is set + module dependency to nb-javac lib is declared?

lahodaj replied:
Sorry for belated reply.

(AFAIK) This will run every time this module is on the appropriate path (and the AP is enabled). I.e. it will run in the editor when the NetBeans modules that depend on this are open (like `java.source.base`). It will not (should not) run for ordinary unrelated user code.

mbien replied:
I have nothing against it but I am just wondering: given that the methods are deprecated and all call sites were updated in this PR (it is also only checking `org/netbeans` paths and nothing external), I am wondering what this is supposed to catch?

Wouldn't grepping the build for compiler warnings related to `TreeMaker` be sufficient?

lahodaj replied:
Sure, grepping for compiler warnings would work. The idea is mostly to make sure we don't re-introduce invocations of the deprecated methods - there are surely various means to achieve that, so it is mostly because it is easier for me to write an AP than a filter that would check for the specific warnings during a build.

mbien replied:
can hint files mark things as error? NB used to use those for migration purposes, e.g:
https://github.com/apache/netbeans/blob/master/platform/openide.filesystems/src/META-INF/upgrade/FileObject.hint

I guess I am wondering why a new way of API usage warnings is introduced now - but maybe I am missing something and the other approaches would not work here. I think it is also fairly unlikely that new code would start calling a deprecated method by accident - but i could be wrong.

mbien replied:
```java
"don't use this":
new Character($l) :: $l instanceof char =>
<!error='blabla'>
;;
```

tried this but it looks like that it won't mark it as error unfortunately. But it marks it as regular warning.

the only usage for the error option is here https://github.com/apache/netbeans/blob/3ebd6cb396842f212bd09011d689941884dd649d/java/java.hints.declarative/src/org/netbeans/modules/java/hints/declarative/DeclarativeHintsWorker.java#L135

mbien replied:
thanks, but i was only asking questions - I was curious why it was implemented that way thats all.

mbien reviewed (approved):
Code looks good to me. Ran a quick manual test and every worked.

Deprecating the method was a good idea - wanted to suggest that too so that it isn't used by accident again.

lahodaj commented:
Any issue with merging this sometime soon?

matthiasblaesing commented:
Given, the broad changes, I think it would be good to get this in early, so I suggest to merge now. That way people running nightlies will test and report problems early.

mbien commented:
i reviewed it and tested it - have no concerns regarding integration.
